
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Getting started with tskit &#8212; SheepTreeSeq</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=87e54e7c" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/01-getting_started';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="TSINFER tutorial" href="02-tsinfer.html" />
    <link rel="prev" title="Quickstart" href="00-quickstart.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="SheepTreeSeq - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="SheepTreeSeq - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    SheepTreeSeq: Scalable analysis of sheep genome diversity using tree sequences
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Setting up the environment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../install/install.html">Setting up the project</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install/build-docs.html">Build the documentation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Introducing TreeSequences</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="00-quickstart.html">Quickstart</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Getting started with tskit</a></li>
<li class="toctree-l1"><a class="reference internal" href="02-tsinfer.html">TSINFER tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="04-tsinfer_with_test_data.html">TSINFER with SMARTER test data</a></li>
<li class="toctree-l1"><a class="reference internal" href="08-test_sample-vcf.html">Inference from a test VCF file</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">The SMARTER database</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../running/smarter-database.html">The SMARTER-database</a></li>
<li class="toctree-l1"><a class="reference internal" href="03-smarter_database.html">Infer ancestry using EST-SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="05-tstree-smarter.html">Reading data from chromosome 26</a></li>
<li class="toctree-l1"><a class="reference internal" href="06-nucleotide-diversity.html">Nucleotide diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="07-about_tstree_space.html">About tstree space</a></li>
<li class="toctree-l1"><a class="reference internal" href="10-ensembl-compara.html">Get ancient information from ensemble compara</a></li>
<li class="toctree-l1"><a class="reference internal" href="09-smarter_50k.html">SMARTER 50K</a></li>
<li class="toctree-l1"><a class="reference internal" href="11-smarter_50K_results.html">50K simulations results</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/bunop/TSKITetude" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/bunop/TSKITetude/issues/new?title=Issue%20on%20page%20%2Fnotebooks/01-getting_started.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/01-getting_started.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Getting started with tskit</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-trees">Processing trees</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-sites-and-mutations">Processing sites and mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-genotypes">Processing genotypes</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="getting-started-with-tskit">
<h1>Getting started with tskit<a class="headerlink" href="#getting-started-with-tskit" title="Link to this heading">#</a></h1>
<p>This is the step-by-step tutorial found <a class="reference external" href="https://tskit.dev/tutorials/getting_started.html">here</a>. Here we generate an alignment using <a class="reference external" href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>, which is a python package to generate data to be used with <em>tskit</em> stuff</p>
<blockquote>
<div><p>A number of different software programs can generate tree sequences. For the purposes of this tutorial we’ll use msprime to create an example tree sequence representing the genetic genealogy of a 10Mb chromosome in twenty diploid individuals. To make it a bit more interesting, we’ll simulate the effects of a selective sweep in the middle of the chromosome, then throw some neutral mutations onto the resulting tree sequence.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">pop_size</span><span class="o">=</span><span class="mi">10_000</span>
<span class="n">seq_length</span><span class="o">=</span><span class="mi">10_000_000</span>

<span class="n">sweep_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">SweepGenicSelection</span><span class="p">(</span>
    <span class="n">position</span><span class="o">=</span><span class="n">seq_length</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">start_frequency</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span> <span class="n">end_frequency</span><span class="o">=</span><span class="mf">0.9999</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">dt</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">)</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">20</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="p">[</span><span class="n">sweep_model</span><span class="p">,</span> <span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">()],</span>
    <span class="n">population_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">seq_length</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span>  <span class="c1"># only needed for repeatabilty</span>
    <span class="p">)</span>
<span class="c1"># Optionally add finite-site mutations to the ts using the Jukes &amp; Cantor model, creating SNPs</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">4321</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>11167</td></tr>
                      <tr><td>Sequence Length</td><td>10000000.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>40</td></tr>
                      <tr><td>Total Size</td><td>2.4 MiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>36372</td>
                      <td>1.1 MiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>20</td>
                      <td>584 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>13568</td>
                      <td>490.3 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>7342</td>
                      <td>200.8 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>2</td>
                      <td>1.9 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>13554</td>
                      <td>330.9 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>We have tousand of trees in <code class="docutils literal notranslate"><span class="pre">ts</span></code> object. We have <em>20 dyploid</em> individuals, so 40 nodes (one for genome? have I <em>two</em> genomes per individual as described by the tutorial?)</p>
<section id="processing-trees">
<h2>Processing trees<a class="headerlink" href="#processing-trees" title="Link to this heading">#</a></h2>
<p>Iterate over the <em>trees</em> with the <code class="docutils literal notranslate"><span class="pre">trees()</span></code> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tree </span><span class="si">{</span><span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="si">}</span><span class="s2"> covers </span><span class="si">{</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">index</span> <span class="o">&gt;=</span> <span class="mi">4</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="k">break</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tree </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">last</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="si">}</span><span class="s2"> covers </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">last</span><span class="p">()</span><span class="o">.</span><span class="n">interval</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Tree 0 covers Interval(left=0.0, right=661.0)
Tree 1 covers Interval(left=661.0, right=3116.0)
Tree 2 covers Interval(left=3116.0, right=4451.0)
Tree 3 covers Interval(left=4451.0, right=4673.0)
Tree 4 covers Interval(left=4673.0, right=5020.0)
...
Tree 11166 covers Interval(left=9999635.0, right=10000000.0)
</pre></div>
</div>
</div>
</div>
<p>There are also <code class="docutils literal notranslate"><span class="pre">last()</span></code> and <code class="docutils literal notranslate"><span class="pre">first()</span></code> methods to access to the <em>last</em> and <em>first</em> trees respectively. Check if trees coalesce (not always true for <a class="reference external" href="https://tskit.dev/tutorials/forward_sims.html#sec-tskit-forward-simulations">forward simulations</a>)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">time</span>
<span class="n">elapsed</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">has_multiple_roots</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Tree </span><span class="si">{tree.index}</span><span class="s2"> has not coalesced&quot;</span><span class="p">)</span>
        <span class="k">break</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">elapsed</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">elapsed</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;All </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span><span class="si">}</span><span class="s2"> trees coalesced&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Checked in </span><span class="si">{</span><span class="n">elapsed</span><span class="si">:</span><span class="s2">.6g</span><span class="si">}</span><span class="s2"> secs&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>All 11167 trees coalesced
Checked in 0.0181918 secs
</pre></div>
</div>
</div>
</div>
<p>Now that we know all trees have coalesced, we know that at each position in the
genome all the 40 sample nodes must have one most recent common ancestor (MRCA).
Below, we iterate over the trees, finding the IDs of the root (MRCA) node for e
ach tree. The time of this root node can be found via the <code class="docutils literal notranslate"><span class="pre">tskit.TreeSequence.node()</span></code>
method, which returns a <code class="docutils literal notranslate"><span class="pre">Node</span></code> object whose attributes include the node time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">kb</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Starting genomic position</span>
<span class="n">mrca_time</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="n">kb</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">right</span><span class="o">/</span><span class="mi">1000</span><span class="p">)</span>  <span class="c1"># convert to kb</span>
    <span class="n">mrca</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">)</span>  <span class="c1"># For msprime tree sequences, the root node is the MRCA</span>
    <span class="n">mrca_time</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrca</span><span class="o">.</span><span class="n">time</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stairs</span><span class="p">(</span><span class="n">mrca_time</span><span class="p">,</span> <span class="n">kb</span><span class="p">,</span> <span class="n">baseline</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Genome position (kb)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Time of root (or MRCA) in generations&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/2c9458cec9811689ef3fc4470e593f4e1a4df79a71589383b17e1b6cfe81611f.png" src="../_images/2c9458cec9811689ef3fc4470e593f4e1a4df79a71589383b17e1b6cfe81611f.png" />
</div>
</div>
<p>It’s obvious that there’s something unusual about the trees in the middle of
this chromosome, where the selective sweep occurred.</p>
<p>Although tskit is designed so that is it rapid to pass through trees sequentially,
it is also possible to pull out individual trees from the middle of a tree sequence
via the <code class="docutils literal notranslate"><span class="pre">TreeSequence.at()</span></code> method. Here’s how you can use that to extract the
tree at location - the position of the sweep - and draw it using the <code class="docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code>
method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">swept_tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">at</span><span class="p">(</span><span class="mi">5_000_000</span><span class="p">)</span>  <span class="c1"># or you can get e.g. the nth tree using ts.at_index(n)</span>
<span class="n">intvl</span> <span class="o">=</span> <span class="n">swept_tree</span><span class="o">.</span><span class="n">interval</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tree number </span><span class="si">{</span><span class="n">swept_tree</span><span class="o">.</span><span class="n">index</span><span class="si">}</span><span class="s2">, which runs from position </span><span class="si">{</span><span class="n">intvl</span><span class="o">.</span><span class="n">left</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">intvl</span><span class="o">.</span><span class="n">right</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
<span class="c1"># Draw it at a wide size, to make room for all 40 tips</span>
<span class="n">swept_tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">200</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Tree number 5382, which runs from position 4998293.0 to 5033047.0:
</pre></div>
</div>
<img alt="../_images/478e935e275f407c117c25ab98589019d31cc0b83876121ed2a1d10f1e4d438a.svg" src="../_images/478e935e275f407c117c25ab98589019d31cc0b83876121ed2a1d10f1e4d438a.svg" />
</div>
</div>
<p>This tree shows the classic signature of a recent expansion or selection event,
with many long terminal branches, resulting in an excess of singleton mutations.</p>
<p>It can often be helpful to slim down a tree sequence so that it represents the
genealogy of a smaller subset of the original samples. This can be done using
the powerful <code class="docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code> method.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">TreeSequence.draw_svg()</span></code> method allows us to draw more than one tree:
either the entire tree sequence, or (by using the <code class="docutils literal notranslate"><span class="pre">x_lim</span></code> parameter) a smaller
region of the genome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reduced_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">])</span>  <span class="c1"># simplify to the first 10 samples</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Genealogy of the first 10 samples for the first 5kb of the genome&quot;</span><span class="p">)</span>
<span class="n">reduced_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">x_lim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5000</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Genealogy of the first 10 samples for the first 5kb of the genome
</pre></div>
</div>
<img alt="../_images/66eb4d7750c4c3158dbd231791de53e5b6ccabbd66daea2190613ab8858f6a88.svg" src="../_images/66eb4d7750c4c3158dbd231791de53e5b6ccabbd66daea2190613ab8858f6a88.svg" />
</div>
</div>
<p>These are much more standard-looking coalescent trees, with far longer branches
higher up in the tree, and therefore many more mutations at higher-frequencies.</p>
<blockquote>
<div><p>In this tutorial we refer to objects, such as sample nodes, by their numerical
IDs. These can change after simplification, and it is often more meaningful
to <a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata">work with metadata</a>,
such as sample and population names, which can be permanently attached to
objects in the tree sequence. Such metadata is often incorporated automatically
by the tools generating the tree sequence.</p>
</div></blockquote>
</section>
<section id="processing-sites-and-mutations">
<h2>Processing sites and mutations<a class="headerlink" href="#processing-sites-and-mutations" title="Link to this heading">#</a></h2>
<p>For many purposes it may be better to focus on the genealogy of your samples,
rather than the sites and mutations that define the genome sequence itself.
Nevertheless, tskit also provides efficient ways to return Site object and Mutation
objects from a tree sequence. For instance, under the finite sites model of mutation
that we used above, multiple mutations can occur at some sites, and we can identify
them by iterating over the sites using the <code class="docutils literal notranslate"><span class="pre">TreeSequence.sites()</span></code> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">num_muts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

<span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
    <span class="n">num_muts</span><span class="p">[</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>  <span class="c1"># site.mutations is a list of mutations at the site</span>

<span class="c1"># Print out some info about mutations per site</span>
<span class="k">for</span> <span class="n">nmuts</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">num_muts</span><span class="p">)):</span>
    <span class="n">info</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> sites&quot;</span>
    <span class="k">if</span> <span class="n">nmuts</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">info</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;, with IDs </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">num_muts</span><span class="o">==</span><span class="n">nmuts</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">,&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;have </span><span class="si">{</span><span class="n">nmuts</span><span class="si">}</span><span class="s2"> mutation&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;s&quot;</span> <span class="k">if</span> <span class="n">nmuts</span> <span class="o">!=</span> <span class="mi">1</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0 sites have 0 mutations
13540 sites have 1 mutation
14 sites, with IDs [ 2275  2287  2658  3789  4043  9694 11023 11780 11853 11890 11999 12023
 12096 12310], have 2 mutations
</pre></div>
</div>
</div>
</div>
</section>
<section id="processing-genotypes">
<h2>Processing genotypes<a class="headerlink" href="#processing-genotypes" title="Link to this heading">#</a></h2>
<p>At each site, the sample nodes will have a particular allelic state (or be flagged
as Missing data). The <code class="docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code> method gives access to the full
variation data. For efficiency, the genotypes at a site are returned as a numpy
array of integers:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">set_printoptions</span><span class="p">(</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>  <span class="c1"># print genotypes on a single line</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Genotypes&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site </span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">genotypes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span> <span class="o">&gt;=</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># only print up to site ID 4</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Genotypes
Site 0: [0 0 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0]
Site 1: [1 1 0 0 1 1 0 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 1 0 1 0 0 1 1]
Site 2: [0 0 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0]
Site 3: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
Site 4: [0 0 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0]
...
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p>Tree sequences are optimised to look at all samples at one site, then all samples
at an adjacent site, and so on along the genome. It is much less efficient look
at all the sites for a single sample, then all the sites for the next sample,
etc. In other words, <strong>you should generally iterate over sites</strong>, not samples.
Nevertheless, all the alleles for a single sample can be obtained via the
<code class="docutils literal notranslate"><span class="pre">TreeSequence.haplotypes()</span></code> method.</p>
</div></blockquote>
<p>To find the actual allelic states at a site, you can refer to the alleles
provided for each Variant: the genotype value is an index into this list.
Here’s one way to print them out; for clarity this example also prints out the IDs
of both the sample nodes (i.e. the genomes) and the diploid individuals in which
each sample node resides.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">samp_ids</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  ID of diploid individual: &quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">individual</span><span class="si">:</span><span class="s2">3</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samp_ids</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;       ID of (sample) node: &quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">s</span><span class="si">:</span><span class="s2">3</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samp_ids</span><span class="p">]))</span>

<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="n">site</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">site</span>
    <span class="n">alleles</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site </span><span class="si">{</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (ancestral state &#39;</span><span class="si">{</span><span class="n">site</span><span class="o">.</span><span class="n">ancestral_state</span><span class="si">}</span><span class="s2">&#39;)&quot;</span><span class="p">,</span>  <span class="n">alleles</span><span class="p">[</span><span class="n">v</span><span class="o">.</span><span class="n">genotypes</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">site</span><span class="o">.</span><span class="n">id</span> <span class="o">&gt;=</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># only print up to site ID 4</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  ID of diploid individual:    0   0   1   1   2   2   3   3   4   4   5   5   6   6   7   7   8   8   9   9  10  10  11  11  12  12  13  13  14  14  15  15  16  16  17  17  18  18  19  19
       ID of (sample) node:    0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36  37  38  39
Site 0 (ancestral state &#39;G&#39;) [&#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39;]
Site 1 (ancestral state &#39;T&#39;) [&#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;T&#39; &#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;G&#39; &#39;G&#39;]
Site 2 (ancestral state &#39;T&#39;) [&#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;C&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
Site 3 (ancestral state &#39;T&#39;) [&#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
Site 4 (ancestral state &#39;T&#39;) [&#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
...
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="00-quickstart.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Quickstart</p>
      </div>
    </a>
    <a class="right-next"
       href="02-tsinfer.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">TSINFER tutorial</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-trees">Processing trees</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-sites-and-mutations">Processing sites and mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#processing-genotypes">Processing genotypes</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Paolo Cozzi
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>