
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>TSINFER with SMARTER test data &#8212; SheepTreeSeq</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!-- 
    this give us a css class that will be invisible only if js is disabled 
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=26a4bc78f4c0ddb94549"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549" />

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/04-tsinfer_with_test_data';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Inference from a test VCF file" href="08-test_sample-vcf.html" />
    <link rel="prev" title="TSINFER tutorial" href="02-tsinfer.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="SheepTreeSeq - Home"/>
    <img src="../_static/logo.png" class="logo__image only-dark pst-js-only" alt="SheepTreeSeq - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    SheepTreeSeq: Scalable analysis of sheep genome diversity using tree sequences
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Setting up the environment</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../install/install.html">Setting up the project</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install/build-docs.html">Build the documentation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Introducing TreeSequences</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="00-quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="01-getting_started.html">Getting started with tskit</a></li>
<li class="toctree-l1"><a class="reference internal" href="02-tsinfer.html">TSINFER tutorial</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">TSINFER with SMARTER test data</a></li>
<li class="toctree-l1"><a class="reference internal" href="08-test_sample-vcf.html">Inference from a test VCF file</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulating TreeSequences</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="14-inspect_simulations.html">Analyze the simulated data</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">The SMARTER database</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../running/smarter-database.html">The SMARTER-database</a></li>
<li class="toctree-l1"><a class="reference internal" href="15-smarter_samples_list.html">Get SMARTER samples list</a></li>
<li class="toctree-l1"><a class="reference internal" href="03-smarter_database.html">Infer ancestry using EST-SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="05-tstree-smarter.html">Reading data from chromosome 26</a></li>
<li class="toctree-l1"><a class="reference internal" href="06-nucleotide-diversity.html">Nucleotide diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="07-about_tstree_space.html">About tstree space</a></li>
<li class="toctree-l1"><a class="reference internal" href="10-ensembl-compara.html">Get ancient information from ensemble compara</a></li>
<li class="toctree-l1"><a class="reference internal" href="09-smarter_50k.html">SMARTER 50K</a></li>
<li class="toctree-l1"><a class="reference internal" href="11-smarter_50K_results.html">50K simulations results</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/bunop/TSKITetude" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/bunop/TSKITetude/issues/new?title=Issue%20on%20page%20%2Fnotebooks/04-tsinfer_with_test_data.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/04-tsinfer_with_test_data.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>TSINFER with SMARTER test data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tsinfer-with-smarter-test-data">
<h1>TSINFER with SMARTER test data<a class="headerlink" href="#tsinfer-with-smarter-test-data" title="Link to this heading">#</a></h1>
<p>I’ve created a test dataset consisting of 10 samples and genotype data for the
chromosome <em>26</em> of the SMARTER database. I’ve created a nextflow pipeline to prepare
a dataset and generate a <em>phased/imputed</em> genotype with <em>Beagle</em>. You can simply test the
pipeline using the nextflow test profile: first collect test input file in the <code class="docutils literal notranslate"><span class="pre">data</span></code> directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>https://github.com/cnr-ibba/nf-treeseq/raw/master/tests/Oar_v3.1_chr26.fna.gz?download<span class="o">=</span><span class="w"> </span>-O<span class="w"> </span>data/Oar_v3.1_chr26.fna.gz
wget<span class="w"> </span>https://raw.githubusercontent.com/cnr-ibba/nf-treeseq/master/tests/test_dataset.tsv<span class="w"> </span>-O<span class="w"> </span>data/test_dataset.tsv
wget<span class="w"> </span>https://raw.githubusercontent.com/cnr-ibba/nf-treeseq/master/tests/test_outgroup.tsv<span class="w"> </span>-O<span class="w"> </span>data/test_outgroup.tsv
wget<span class="w"> </span>https://github.com/cnr-ibba/nf-treeseq/raw/master/tests/test_dataset.bed?download<span class="o">=</span><span class="w"> </span>-O<span class="w"> </span>data/test_dataset.bed
wget<span class="w"> </span>https://github.com/cnr-ibba/nf-treeseq/raw/master/tests/test_dataset.bim?download<span class="o">=</span><span class="w"> </span>-O<span class="w"> </span>data/test_dataset.bim
wget<span class="w"> </span>https://github.com/cnr-ibba/nf-treeseq/raw/master/tests/test_dataset.fam?download<span class="o">=</span><span class="w"> </span>-O<span class="w"> </span>data/test_dataset.fam
</pre></div>
</div>
<p>Then run the pipeline with the test profile:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>cnr-ibba/nf-treeseq<span class="w"> </span>-r<span class="w"> </span>v0.2.1<span class="w"> </span>-profile<span class="w"> </span>test,singularity<span class="w"> </span>--plink_bfile<span class="w"> </span>data/test_dataset<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--plink_keep<span class="w"> </span>data/test_dataset.tsv<span class="w"> </span>--genome<span class="w"> </span>data/Oar_v3.1_chr26.fna.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--outdir<span class="w"> </span>results-estsfs/test<span class="w"> </span>--with_estsfs<span class="w"> </span>--outgroup1<span class="w"> </span>data/test_outgroup.tsv
</pre></div>
</div>
<p>Now try to read and determine a <em>tstree</em> object with <em>tsdata</em>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">json</span>

<span class="kn">import</span> <span class="nn">tsinfer</span>
<span class="kn">import</span> <span class="nn">tsdate</span>
<span class="kn">import</span> <span class="nn">cyvcf2</span>
<span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">tskit</span> <span class="kn">import</span> <span class="n">MISSING_DATA</span>

<span class="kn">from</span> <span class="nn">tskitetude</span> <span class="kn">import</span> <span class="n">get_project_dir</span>
<span class="kn">from</span> <span class="nn">tskitetude.helper</span> <span class="kn">import</span> <span class="n">add_populations</span><span class="p">,</span> <span class="n">add_diploid_individuals</span><span class="p">,</span> <span class="n">get_ancestors_alleles</span>
</pre></div>
</div>
</div>
</div>
<p>Define some useful stuff:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_chromosome_lengths</span><span class="p">(</span><span class="n">vcf</span><span class="p">):</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">seqname</span><span class="p">,</span> <span class="n">seqlen</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">seqnames</span><span class="p">,</span> <span class="n">vcf</span><span class="o">.</span><span class="n">seqlens</span><span class="p">):</span>
        <span class="n">results</span><span class="p">[</span><span class="n">seqname</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqlen</span>

    <span class="k">return</span> <span class="n">results</span>

<span class="n">vcf_location</span> <span class="o">=</span> <span class="n">get_project_dir</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;results-estsfs/test/focal/test_dataset.focal.26.vcf.gz&quot;</span>
<span class="n">samples_location</span> <span class="o">=</span> <span class="n">get_project_dir</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;results-estsfs/test/tsinfer/test_dataset.focal.26.samples&quot;</span>

<span class="n">vcf</span> <span class="o">=</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">vcf_location</span><span class="p">)</span>
<span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="n">get_chromosome_lengths</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>I’ve derived ancient alleles with <code class="docutils literal notranslate"><span class="pre">est-sfs</span></code>. Try to load data from my results:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ancestors_alleles</span> <span class="o">=</span> <span class="n">get_ancestors_alleles</span><span class="p">(</span><span class="n">get_project_dir</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;results-estsfs/test/estsfs/samples-merged.26.ancestral.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now try to define a custom function to deal with VCF data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">add_diploid_sites</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">ancestors_alleles</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the sites in the vcf and add them to the samples object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># You may want to change the following line, e.g. here we allow</span>
    <span class="c1"># &quot;*&quot; (a spanning deletion) to be a valid allele state</span>
    <span class="n">allele_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;ATGCatgc*&quot;</span><span class="p">)</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">progressbar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">samples</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Read VCF&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s1">&#39;bp&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>  <span class="c1"># Loop over variants, each assumed at a unique site</span>
        <span class="n">progressbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span> <span class="o">-</span> <span class="n">pos</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">pos</span> <span class="o">==</span> <span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Duplicate entries at position </span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">, ignoring all but the first&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">POS</span>

        <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="ow">not</span> <span class="n">phased</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">phased</span> <span class="ow">in</span> <span class="n">variant</span><span class="o">.</span><span class="n">genotypes</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unphased genotypes for variant at position&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>

        <span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">REF</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span> <span class="o">+</span> <span class="p">[</span><span class="n">v</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">variant</span><span class="o">.</span><span class="n">ALT</span><span class="p">]</span>
        <span class="n">ancestral_allele</span> <span class="o">=</span> <span class="n">ancestors_alleles</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="n">variant</span><span class="o">.</span><span class="n">CHROM</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="p">),</span> <span class="n">MISSING_DATA</span><span class="p">)</span>

        <span class="c1"># Check we have ATCG alleles</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="o">-</span> <span class="n">allele_chars</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ignoring site at pos </span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">: allele </span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s2"> not in </span><span class="si">{</span><span class="n">allele_chars</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

        <span class="c1"># Map original allele indexes to their indexes in the new alleles list.</span>
        <span class="n">genotypes</span> <span class="o">=</span> <span class="p">[</span><span class="n">g</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">variant</span><span class="o">.</span><span class="n">genotypes</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]]</span>
        <span class="n">samples</span><span class="o">.</span><span class="n">add_site</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">,</span> <span class="n">alleles</span><span class="p">,</span> <span class="n">ancestral_allele</span><span class="o">=</span><span class="n">ancestral_allele</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Add individual and populations to empty samples data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">tsinfer</span><span class="o">.</span><span class="n">SampleData</span><span class="p">(</span>
        <span class="n">path</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">samples_location</span><span class="p">),</span> <span class="n">sequence_length</span><span class="o">=</span><span class="n">chromosome_lengths</span><span class="p">[</span><span class="s2">&quot;26&quot;</span><span class="p">])</span> <span class="k">as</span> <span class="n">samples</span><span class="p">:</span>
    <span class="n">samples_tsv</span> <span class="o">=</span> <span class="n">get_project_dir</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;data/test_dataset.tsv&quot;</span>
    <span class="n">pop_lookup</span> <span class="o">=</span> <span class="n">add_populations</span><span class="p">(</span><span class="n">samples_tsv</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
    <span class="n">indv_lookup</span> <span class="o">=</span> <span class="n">add_diploid_individuals</span><span class="p">(</span><span class="n">samples_tsv</span><span class="p">,</span> <span class="n">pop_lookup</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
    <span class="n">add_diploid_sites</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">ancestors_alleles</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/cozzip/.cache/pypoetry/virtualenvs/tskitetude-hh-GIRXc-py3.10/lib/python3.10/site-packages/tsinfer/formats.py:458: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  return zarr.LMDBStore(self.path, subdir=False, map_size=map_size)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "6da06d51c42f48bc97656d6f3bbc309f"}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/cozzip/.cache/pypoetry/virtualenvs/tskitetude-hh-GIRXc-py3.10/lib/python3.10/site-packages/tsinfer/formats.py:428: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  store = zarr.LMDBStore(
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Sample file created for </span><span class="si">{}</span><span class="s2"> samples &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span>
    <span class="o">+</span> <span class="s2">&quot;(</span><span class="si">{}</span><span class="s2"> individuals) &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">num_individuals</span><span class="p">)</span>
    <span class="o">+</span> <span class="s2">&quot;with </span><span class="si">{}</span><span class="s2"> variable sites.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">num_sites</span><span class="p">),</span>
    <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Do the inference</span>
<span class="n">sparrow_ts</span> <span class="o">=</span> <span class="n">tsinfer</span><span class="o">.</span><span class="n">infer</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Inferred tree sequence `</span><span class="si">{}</span><span class="s2">`: </span><span class="si">{}</span><span class="s2"> trees over </span><span class="si">{}</span><span class="s2"> Mb&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="s2">&quot;sparrow_ts&quot;</span><span class="p">,</span> <span class="n">sparrow_ts</span><span class="o">.</span><span class="n">num_trees</span><span class="p">,</span> <span class="n">sparrow_ts</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">/</span> <span class="mf">1e6</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="c1"># Check the metadata</span>
<span class="k">for</span> <span class="n">sample_node_id</span> <span class="ow">in</span> <span class="n">sparrow_ts</span><span class="o">.</span><span class="n">samples</span><span class="p">():</span>
    <span class="n">individual_id</span> <span class="o">=</span> <span class="n">sparrow_ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">sample_node_id</span><span class="p">)</span><span class="o">.</span><span class="n">individual</span>
    <span class="n">population_id</span> <span class="o">=</span> <span class="n">sparrow_ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">sample_node_id</span><span class="p">)</span><span class="o">.</span><span class="n">population</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Node&quot;</span><span class="p">,</span>
        <span class="n">sample_node_id</span><span class="p">,</span>
        <span class="s2">&quot;labels a chr26 sampled from individual&quot;</span><span class="p">,</span>
        <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">sparrow_ts</span><span class="o">.</span><span class="n">individual</span><span class="p">(</span><span class="n">individual_id</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">),</span>
        <span class="s2">&quot;in&quot;</span><span class="p">,</span>
        <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">sparrow_ts</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">population_id</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">),</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample file created for 18 samples (9 individuals) with 680 variable sites.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,492 - tsinfer.inference - INFO - Starting addition of 680 sites
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,524 - tsinfer.inference - INFO - Finished adding sites
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,525 - tsinfer.inference - INFO - Starting build for 525 ancestors
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,640 - tsinfer.inference - INFO - Finished building ancestors
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,646 - tsinfer.inference - INFO - Mismatch prevented by setting constant high recombination and low mismatch probabilities
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,647 - tsinfer.inference - INFO - Summary of recombination probabilities between sites: min=0.01; max=0.01; median=0.01; mean=0.01
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,649 - tsinfer.inference - INFO - Summary of mismatch probabilities over sites: min=1e-20; max=1e-20; median=1e-20; mean=1e-20
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,649 - tsinfer.inference - INFO - Matching using 13 digits of precision in likelihood calcs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,659 - tsinfer.inference - INFO - Starting ancestor matching for 18 epochs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,748 - tsinfer.inference - INFO - Built ancestors tree sequence: 591 nodes (64 pc ancestors); 2587 edges; 527 sites; 527 mutations
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,749 - tsinfer.inference - INFO - Finished ancestor matching
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,754 - tsinfer.inference - INFO - Mismatch prevented by setting constant high recombination and low mismatch probabilities
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,755 - tsinfer.inference - INFO - Summary of recombination probabilities between sites: min=0.01; max=0.01; median=0.01; mean=0.01
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,757 - tsinfer.inference - INFO - Summary of mismatch probabilities over sites: min=1e-20; max=1e-20; median=1e-20; mean=1e-20
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,758 - tsinfer.inference - INFO - Matching using 13 digits of precision in likelihood calcs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,774 - tsinfer.inference - INFO - Loaded 18 samples 591 nodes; 2587 edges; 527 sites; 527 mutations
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,775 - tsinfer.inference - INFO - Started matching for 18 samples
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,801 - tsinfer.inference - INFO - Inserting sample paths: 1163 edges in total
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,805 - tsinfer.inference - INFO - Finalising tree sequence
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,826 - tsinfer.inference - INFO - Built samples tree sequence: 617 nodes (72 pc); 3746 edges; 527 sites; 527 mutations
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,828 - tsinfer.inference - INFO - Mapping additional sites
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,854 - tsinfer.inference - INFO - Removing the oldest edge to detach the virtual-root-like ancestor
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,860 - tsinfer.inference - INFO - Located the all zeros ultimate ancestor
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,864 - tsinfer.inference - INFO - Splitting ultimate ancestor into 276 nodes
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,881 - tsinfer.inference - INFO - Erased flanks covering 0.5445215377117799% of the genome: 166516.0 units at the start and 73497.0 units at the end
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,882 - tsinfer.inference - INFO - Simplifying with filter_sites=False, filter_populations=False, filter_individuals=False, and keep_unary=True on 893 nodes and 5212 edges
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,884 - tsinfer.inference - INFO - Finished simplify; now have 886 nodes and 3559 edges
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Inferred tree sequence `sparrow_ts`: 513 trees over 44.077779 Mb
Node 0 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-TEX-000000001&#39;} in {&#39;breed&#39;: &#39;TEX&#39;}
Node 1 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-TEX-000000001&#39;} in {&#39;breed&#39;: &#39;TEX&#39;}
Node 2 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;GROA-FRZ-000000170&#39;} in {&#39;breed&#39;: &#39;FRZ&#39;}
Node 3 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;GROA-FRZ-000000170&#39;} in {&#39;breed&#39;: &#39;FRZ&#39;}
Node 4 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-MER-000000224&#39;} in {&#39;breed&#39;: &#39;MER&#39;}
Node 5 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-MER-000000224&#39;} in {&#39;breed&#39;: &#39;MER&#39;}
Node 6 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-CRR-000000320&#39;} in {&#39;breed&#39;: &#39;CRR&#39;}
Node 7 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-CRR-000000320&#39;} in {&#39;breed&#39;: &#39;CRR&#39;}
Node 8 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-CRL-000000380&#39;} in {&#39;breed&#39;: &#39;CRL&#39;}
Node 9 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;UYOA-CRL-000000380&#39;} in {&#39;breed&#39;: &#39;CRL&#39;}
Node 10 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;FROA-BER-000000478&#39;} in {&#39;breed&#39;: &#39;BER&#39;}
Node 11 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;FROA-BER-000000478&#39;} in {&#39;breed&#39;: &#39;BER&#39;}
Node 12 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-ALT-000001084&#39;} in {&#39;breed&#39;: &#39;ALT&#39;}
Node 13 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-ALT-000001084&#39;} in {&#39;breed&#39;: &#39;ALT&#39;}
Node 14 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-LEC-000002620&#39;} in {&#39;breed&#39;: &#39;LEC&#39;}
Node 15 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-LEC-000002620&#39;} in {&#39;breed&#39;: &#39;LEC&#39;}
Node 16 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-SAB-000003324&#39;} in {&#39;breed&#39;: &#39;SAB&#39;}
Node 17 labels a chr26 sampled from individual {&#39;sample_id&#39;: &#39;ITOA-SAB-000003324&#39;} in {&#39;breed&#39;: &#39;SAB&#39;}
</pre></div>
</div>
</div>
</div>
<p>Try to infer <em>dates</em> on my tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Removes unary nodes (currently required in tsdate), keeps historical-only sites</span>
<span class="n">inferred_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">preprocess_ts</span><span class="p">(</span><span class="n">sparrow_ts</span><span class="p">,</span> <span class="n">filter_sites</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">dated_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">date</span><span class="p">(</span><span class="n">inferred_ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">Ne</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>

<span class="n">dated_ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,899 - tsdate.util - INFO - Beginning preprocessing
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,900 - tsdate.util - INFO - Minimum_gap: None and remove_telomeres: None
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,901 - tsdate.util - INFO - REMOVING TELOMERE: Snip topology from 0 to first site at 166515.0.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-11-12 18:07:28,902 - tsdate.util - INFO - REMOVING TELOMERE: Snip topology from 44004282.0 to end of sequence at 44077779.0.
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">TypeError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">7</span><span class="p">],</span> <span class="n">line</span> <span class="mi">3</span>
<span class="g g-Whitespace">      </span><span class="mi">1</span> <span class="c1"># Removes unary nodes (currently required in tsdate), keeps historical-only sites</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="n">inferred_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">preprocess_ts</span><span class="p">(</span><span class="n">sparrow_ts</span><span class="p">,</span> <span class="n">filter_sites</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">3</span> <span class="n">dated_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">date</span><span class="p">(</span><span class="n">inferred_ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">Ne</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> <span class="n">dated_ts</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/tskitetude-hh-GIRXc-py3.10/lib/python3.10/site-packages/tsdate/core.py:1751,</span> in <span class="ni">date</span><span class="nt">(tree_sequence, mutation_rate, recombination_rate, time_units, method, constr_iterations, return_posteriors, return_likelihood, progress, record_provenance, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">1748</span> <span class="k">if</span> <span class="n">method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">estimation_methods</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1749</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;method must be one of </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">estimation_methods</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1751</span> <span class="k">return</span> <span class="n">estimation_methods</span><span class="p">[</span><span class="n">method</span><span class="p">](</span>
<span class="g g-Whitespace">   </span><span class="mi">1752</span>     <span class="n">tree_sequence</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1753</span>     <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1754</span>     <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1755</span>     <span class="n">time_units</span><span class="o">=</span><span class="n">time_units</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1756</span>     <span class="n">progress</span><span class="o">=</span><span class="n">progress</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1757</span>     <span class="n">constr_iterations</span><span class="o">=</span><span class="n">constr_iterations</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1758</span>     <span class="n">return_posteriors</span><span class="o">=</span><span class="n">return_posteriors</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1759</span>     <span class="n">return_likelihood</span><span class="o">=</span><span class="n">return_likelihood</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1760</span>     <span class="n">record_provenance</span><span class="o">=</span><span class="n">record_provenance</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1761</span>     <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1762</span> <span class="p">)</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/tskitetude-hh-GIRXc-py3.10/lib/python3.10/site-packages/tsdate/core.py:1634,</span> in <span class="ni">variational_gamma</span><span class="nt">(tree_sequence, mutation_rate, eps, max_iterations, rescaling_intervals, rescaling_iterations, match_segregating_sites, max_shape, regularise_roots, singletons_phased, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">1630</span> <span class="k">if</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">num_mutations</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1631</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1632</span>         <span class="s2">&quot;No mutations present: these are required for the variational_gamma method&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1633</span>     <span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1634</span> <span class="n">dating_method</span> <span class="o">=</span> <span class="n">VariationalGammaMethod</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1635</span>     <span class="n">tree_sequence</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="g g-Whitespace">   </span><span class="mi">1636</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1637</span> <span class="n">result</span> <span class="o">=</span> <span class="n">dating_method</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1638</span>     <span class="n">eps</span><span class="o">=</span><span class="n">eps</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1639</span>     <span class="n">max_iterations</span><span class="o">=</span><span class="n">max_iterations</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1645</span>     <span class="n">singletons_phased</span><span class="o">=</span><span class="n">singletons_phased</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1646</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1647</span> <span class="k">return</span> <span class="n">dating_method</span><span class="o">.</span><span class="n">parse_result</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">eps</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;parameter&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;shape&quot;</span><span class="p">,</span> <span class="s2">&quot;rate&quot;</span><span class="p">]})</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/tskitetude-hh-GIRXc-py3.10/lib/python3.10/site-packages/tsdate/core.py:1218,</span> in <span class="ni">VariationalGammaMethod.__init__</span><span class="nt">(self, ts, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">1217</span> <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="ne">-&gt; </span><span class="mi">1218</span>     <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="ne">TypeError</span>: EstimationMethod.__init__() got an unexpected keyword argument &#39;Ne&#39;
</pre></div>
</div>
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="02-tsinfer.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">TSINFER tutorial</p>
      </div>
    </a>
    <a class="right-next"
       href="08-test_sample-vcf.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Inference from a test VCF file</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Paolo Cozzi
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>