{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nucleotide diversity\n",
    "\n",
    "## Using REF as ancestral alleles\n",
    "\n",
    "We are wondering how the ancestral allele inference effect the final treesequence object.\n",
    "We are testing a *tstree* object created by imposing the reference allele as the \n",
    "ancestral allele. The effect is that there are more or less the same mutations as the\n",
    "number of variants (while in the case of the ancestrall allele calculated with \n",
    "`est-sfs` we can see million of mutations). Let's start by loading data for this\n",
    "*REF-based* treeseq object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import tskit\n",
    "\n",
    "from tskitetude import get_project_dir\n",
    "from tskitetude.helper import create_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = tskit.load(str(get_project_dir() / \"experiments/smarter-background-sheeps/SMARTER-OA-OAR3-forward-0.4.9.focal.26.trees\"))\n",
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I need to calculate nucleotide diversity *per site*. The only way to do this seems\n",
    "to be calculating windows containing the SNPs an then calculating the nucleotide\n",
    "diversity with the `tskit.TreeSequence.diversity` function. I've created a function\n",
    "`create_windows` in `helper` module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the last index is a simply a 2 step starting from position 1\n",
    "ts_diversity = ts.diversity(windows=create_windows(ts))[1::2]\n",
    "ts_diversity[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's compare the nucleotide diversity calculated using vcftools: here's the \n",
    "command line to calculate nucleotide diversity *per site*:\n",
    "\n",
    "```bash\n",
    "cd experiments/smarter-background-sheeps/\n",
    "vcftools --gzvcf SMARTER-OA-OAR3-forward-0.4.9.focal.26.vcf.gz --out allsamples_pi --site-pi\n",
    "```\n",
    "\n",
    "The `allsamples_pi.sites.pi` is a *TSV* file with the positions and the nucleotide diversity. Read it with pandas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools_diversity = pd.read_csv(get_project_dir() / \"experiments/smarter-background-sheeps/allsamples_pi.sites.pi\", sep=\"\\t\")\n",
    "vcftools_diversity.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are this values similar?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.isclose(ts_diversity, vcftools_diversity[\"PI\"], atol=1e-6).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate diversity using *branch*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the last index is a simply a 2 step starting from position 1\n",
    "ts_diversity_branch = ts.diversity(mode='branch', windows=create_windows(ts))[1::2]\n",
    "ts_diversity_branch[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Try to plot the tow different diversities with vcftools output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(ts_diversity, vcftools_diversity[\"PI\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(ts_diversity_branch, vcftools_diversity[\"PI\"])\n",
    "plt.xlim(0, 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EST-SFS output as ancestral alleles\n",
    "\n",
    "Can we calculate nucleotide diversity using the *tree files* generated by the pipeline\n",
    "using the `est-sfs` output as ancestral alleles?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts1 = tskit.load(str(get_project_dir() / \"results/tsinfer/SMARTER-OA-OAR3-forward-0.4.9.focal.26.trees\"))\n",
    "ts1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Same stuff as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the last index is a simply a 2 step starting from position 1\n",
    "ts1_diversity = ts1.diversity(windows=create_windows(ts1))[1::2]\n",
    "ts1_diversity[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are this values similar to the values calculated using VCFtools?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.isclose(ts1_diversity, vcftools_diversity[\"PI\"], atol=1e-6).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So *nucleotide diversity* is the same in both cases (using the REF as ancestral allele and using the `est-sfs` output as ancestral allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
