{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f467f90d",
   "metadata": {},
   "source": [
    "# Dataset for TS inference\n",
    "\n",
    "Diploid, phased VCF files with varying numbers of individuals. All are BGZIP-compressed and have a CSI index. A key to the populations is in the file `popKey`.\n",
    "\n",
    "These are for TS inference with SINGER, threads, and tskinfer/tsdate. For each file, the REF allele can be assumed to be ancestral. The imputed files with Maj in their name have the REF allele set to the major allele. This is to assess the effect of wrongly specified ancestral alleles.\n",
    "\n",
    "- **Complete VCF**\n",
    "  - `tsm100M300I.vcf.gz` (exported from msprime with ref allele = ancestral)\n",
    "- **Array-density files**\n",
    "  - `ts300I2k.vcf.gz`\n",
    "  - `ts300I25k.vcf.gz`\n",
    "- **Imputed datasets for TS inference**\n",
    "  - `panelLarge25kImputed.vcf.gz`\n",
    "  - `panelLarge25kMajImputed.vcf.gz`\n",
    "  - `panelLarge2kImputed.vcf.gz`\n",
    "  - `panelLarge2kMajImputed.vcf.gz`\n",
    "  - `panelNoBLarge25kImputed.vcf.gz`\n",
    "  - `panelNoBLarge25kMajImputed.vcf.gz`\n",
    "  - `panelNoBLarge2kImputed.vcf.gz`\n",
    "  - `panelNoBLarge2kMajImputed.vcf.gz`\n",
    "  - `panelSmall25kImputed.vcf.gz`\n",
    "  - `panelSmall25kMajImputed.vcf.gz`\n",
    "  - `panelSmall2kImputed.vcf.gz`\n",
    "  - `panelSmall2kMajImputed.vcf.gz`\n",
    "  - `panelNoBSmall25kImputed.vcf.gz`\n",
    "  - `panelNoBSmall25kMajImputed.vcf.gz`\n",
    "  - `panelNoBSmall2kImputed.vcf.gz`\n",
    "  - `panelNoBSmall2kMajImputed.vcf.gz`\n",
    "\n",
    "## Fix file names\n",
    "\n",
    "Files has beed downloaded and extracted in `data/toInfer/` folder: there's a \n",
    "repetition of \"vcf\" in the file names. You can fix this by creating a simple script \n",
    "like this in the `data/toInfer/` folder:\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "# replace .vcf.vcf extensions with .vcf\n",
    "# ex. panelLarge25kImputed.vcf.vcf.gz -> panelLarge25kImputed.vcf.gz\n",
    "set -euo pipefail\n",
    "\n",
    "for file in $(ls *.vcf.vcf.*); do\n",
    "    newname=$(echo \"$file\" | sed 's/\\.vcf\\.vcf/\\.vcf/')\n",
    "    mv \"$file\" \"$newname\"\n",
    "done\n",
    "```\n",
    "\n",
    "Make the script executable with `chmod +x fix_filenames.sh` and run it with \n",
    "`./fix_filenames.sh`.\n",
    "\n",
    "## Create a TSV file sample information\n",
    "\n",
    "Those files are a simulation of 1600 individuals from the original 2405 individuals\n",
    "in the msprime simulation. We need to create a TSV file with sample information\n",
    "using `FID` and `IID` columns: you can create it with an utility script in this\n",
    "folder:\n",
    "\n",
    "```bash\n",
    "python scripts/createFID-IID.py \\\n",
    "    --indiv-list data/toInfer/popKey \\\n",
    "    --directory data/toInfer\n",
    "```\n",
    "\n",
    "## Create a fake FASTA file\n",
    "\n",
    "We need to create a fake FASTA in order to explot `bcftools reheader` to add\n",
    "contig length information to the VCF files: this is required by `create_tstree`\n",
    "script installed in this project (and to be used by the <https://github.com/cnr-ibba/nf-treeseq>\n",
    "Nextflow pipeline). This can be done again with an utility script in this folder:\n",
    "\n",
    "```bash\n",
    "python scripts/fakeFastaFromVCF.py \\\n",
    "    --vcf data/toInfer/tsm100M300I.vcf.gz \\\n",
    "    --output data/toInfer/tsm100M300I.fa.gz\n",
    "```\n",
    "\n",
    "## Call nextflow pipeline\n",
    "\n",
    "You can create two nextflow configuration files for running the `cnr-ibba/nf-treeseq`\n",
    "pipeline: see `config/samples_toInfer.csv` and `config/samples_toInfer.json` to\n",
    "see how to set it up. Then you can call the pipeline with:\n",
    "\n",
    "```bash\n",
    "nextflow run cnr-ibba/nf-treeseq -r issue-8 \\\n",
    "    -profile ibba,core -resume \\\n",
    "    -params-file config/samples_toInfer.json\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccaf626c",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
