{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cf10d687",
   "metadata": {},
   "source": [
    "# Explore threads metadata\n",
    "\n",
    "Threads metadata has been added to the threads TreeSequence with the `annotate_tree`\n",
    "script. Let's explore data and check if everything is fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a6f9fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "import cyvcf2\n",
    "import tszip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from tskitetude import get_data_dir, get_project_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0533377",
   "metadata": {},
   "source": [
    "let's load information on sample names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b7d623fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "population",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "individual",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "f4d44e05-0c43-4fd9-883b-2cecb6c7131c",
       "rows": [
        [
         "0",
         "A",
         "tsk_486"
        ],
        [
         "1",
         "II",
         "tsk_276"
        ],
        [
         "2",
         "A",
         "tsk_506"
        ],
        [
         "3",
         "II",
         "tsk_115"
        ],
        [
         "4",
         "D",
         "tsk_1388"
        ]
       ],
       "shape": {
        "columns": 2,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>individual</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>tsk_486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>II</td>\n",
       "      <td>tsk_276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>tsk_506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>II</td>\n",
       "      <td>tsk_115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D</td>\n",
       "      <td>tsk_1388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population individual\n",
       "0          A    tsk_486\n",
       "1         II    tsk_276\n",
       "2          A    tsk_506\n",
       "3         II    tsk_115\n",
       "4          D   tsk_1388"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info = pd.read_csv(\n",
    "    get_data_dir() / \"toInfer/tsm100M300I.sample_names.txt\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=[\"population\", \"individual\"]\n",
    ")\n",
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c90f6fc",
   "metadata": {},
   "source": [
    "Now open the threads object:\n",
    "\n",
    ":::{warning}\n",
    "where are opening the fitted version of the trees, since they contain variant infomation\n",
    ":::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2aca2941",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "            </style>\n",
       "            <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"padding:0;line-height:21px;\">\n",
       "                                    <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                                    <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                                </th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            <tr><td>Trees</td><td>1 999</td></tr>\n",
       "                            <tr><td>Sequence Length</td><td>99 981 261</td></tr>\n",
       "                            <tr><td>Time Units</td><td>unknown</td></tr>\n",
       "                            <tr><td>Sample Nodes</td><td>3 200</td></tr>\n",
       "                            <tr><td>Total Size</td><td>85.3 MiB</td></tr>\n",
       "                            <tr>\n",
       "                                <td>Metadata</td><td style=\"text-align: left;\">\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\"></span>\n",
       "                <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    chromosome: 1<br/>offset: 8568<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            </td>\n",
       "                            </tr>\n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th style=\"line-height:21px;\">Table</th>\n",
       "                                <th>Rows</th>\n",
       "                                <th>Size</th>\n",
       "                                <th>Has Metadata</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "            <tr>\n",
       "                <td>Edges</td>\n",
       "                <td>1 979 642</td>\n",
       "                <td>60.4 MiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Individuals</td>\n",
       "                <td>1 600</td>\n",
       "                <td>80.8 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    ✅\n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Migrations</td>\n",
       "                <td>0</td>\n",
       "                <td>8 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Mutations</td>\n",
       "                <td>1 998</td>\n",
       "                <td>72.2 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Nodes</td>\n",
       "                <td>358 846</td>\n",
       "                <td>9.6 MiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Populations</td>\n",
       "                <td>8</td>\n",
       "                <td>271 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    ✅\n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Provenances</td>\n",
       "                <td>1</td>\n",
       "                <td>384 Bytes</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <td>Sites</td>\n",
       "                <td>1 998</td>\n",
       "                <td>48.8 KiB</td>\n",
       "                <td style=\"text-align: center;\">\n",
       "                    \n",
       "                </td>\n",
       "            </tr>\n",
       "        \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                    <table class=\"tskit-table\">\n",
       "                        <thead>\n",
       "                            <tr>\n",
       "                                <th>Provenance Timestamp</th>\n",
       "                                <th>Software Name</th>\n",
       "                                <th>Version</th>\n",
       "                                <th>Command</th>\n",
       "                                <th>Full record</th>\n",
       "                            </tr>\n",
       "                        </thead>\n",
       "                        <tbody>\n",
       "                            \n",
       "                <tr>\n",
       "                    <td>09 December, 2025 at 12:13:23 PM</td>\n",
       "                    <td>threads</td>\n",
       "                    <td>0.2.1</td>\n",
       "                    <td>Unknown</td>\n",
       "                    <td>\n",
       "                        <details>\n",
       "                            <summary>Details</summary>\n",
       "                            \n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\"></span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">software:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    name: threads<br/>version: 0.2.1<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>\n",
       "            <div>\n",
       "                <span class=\"tskit-details-label\">parameters:</span>\n",
       "                <details >\n",
       "                    <summary>dict</summary>\n",
       "                    input_file: ts300I2k.1.biallelic.vcf.gz<br/>metadata_added: True<br/>populations_added: 8<br/>individuals_added: 1600<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            <br/>timestamp: 2025-12-<br/>09T12:13:23.094293+00:00<br/>description: Tree sequence annotated with<br/>population and individual<br/>metadata<br/>\n",
       "                </details>\n",
       "            </div>\n",
       "            \n",
       "                        </details>\n",
       "                    </td>\n",
       "                </tr>\n",
       "            \n",
       "                        </tbody>\n",
       "                    </table>\n",
       "                </div>\n",
       "            </div>\n",
       "            <div style=\"margin-top: 1em; font-size: 0.8em; text-align: center;\">\n",
       "                To cite this software, please consult the citation manual: <a href=\"https://tskit.dev/citation/\" target=\"_blank\">https://tskit.dev/citation/</a>\n",
       "            </div>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fdd0a998b00>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_threads = tszip.load(get_project_dir() / \"results-threads/toInfer-fit/threads/ts300I2k.1.trees.tsz\")\n",
    "ts_threads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b072dbc",
   "metadata": {},
   "source": [
    "Explore the metadata of the first 10 nodes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cd79504d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node 0: population=Population(id=7, metadata={'breed': 'II'}), individual=Individual(id=0, flags=0, location=array([], dtype=float64), parents=array([], dtype=int32), nodes=array([0, 1], dtype=int32), metadata={'sample_id': 'tsk_105'})\n",
      "Node 1: population=Population(id=7, metadata={'breed': 'II'}), individual=Individual(id=0, flags=0, location=array([], dtype=float64), parents=array([], dtype=int32), nodes=array([0, 1], dtype=int32), metadata={'sample_id': 'tsk_105'})\n",
      "Node 2: population=Population(id=7, metadata={'breed': 'II'}), individual=Individual(id=1, flags=0, location=array([], dtype=float64), parents=array([], dtype=int32), nodes=array([2, 3], dtype=int32), metadata={'sample_id': 'tsk_106'})\n",
      "Node 3: population=Population(id=7, metadata={'breed': 'II'}), individual=Individual(id=1, flags=0, location=array([], dtype=float64), parents=array([], dtype=int32), nodes=array([2, 3], dtype=int32), metadata={'sample_id': 'tsk_106'})\n",
      "Node 4: population=Population(id=7, metadata={'breed': 'II'}), individual=Individual(id=2, flags=0, location=array([], dtype=float64), parents=array([], dtype=int32), nodes=array([4, 5], dtype=int32), metadata={'sample_id': 'tsk_107'})\n"
     ]
    }
   ],
   "source": [
    "for node in itertools.islice(ts_threads.nodes(), 5):\n",
    "    print(f\"Node {node.id}: population={ts_threads.population(node.population)}, individual={ts_threads.individual(node.individual)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d3fa3a",
   "metadata": {},
   "source": [
    "Ok, seems that metadata follow the sample order in vcf file. Let's prove it by checking\n",
    "the sample names in `ts_threads.individuals()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2fbc4407",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. samples: 3200\n",
      "Num. individuals: 1600\n",
      "Individual 0: sample='tsk_105', nodes=[0, 1]\n",
      "Individual 1: sample='tsk_106', nodes=[2, 3]\n",
      "Individual 2: sample='tsk_107', nodes=[4, 5]\n",
      "Individual 3: sample='tsk_108', nodes=[6, 7]\n",
      "Individual 4: sample='tsk_109', nodes=[8, 9]\n"
     ]
    }
   ],
   "source": [
    "print(f\"Num. samples: {ts_threads.num_samples}\")\n",
    "print(f\"Num. individuals: {ts_threads.num_individuals}\")\n",
    "\n",
    "sample2nodes = {}\n",
    "\n",
    "for i, individual in enumerate(ts_threads.individuals()):\n",
    "    if i < 5:\n",
    "        print(f\"Individual {individual.id}: sample='{individual.metadata[\"sample_id\"]}', nodes={individual.nodes.tolist()}\")\n",
    "\n",
    "    sample2nodes[individual.metadata[\"sample_id\"]] = individual.nodes.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4091883",
   "metadata": {},
   "source": [
    "Let's discover individual breeds using their node informations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5a1d8d38",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample2breed = {}\n",
    "\n",
    "for sample_id, nodes in sample2nodes.items():\n",
    "    for node in nodes:\n",
    "        sample2breed.setdefault(\n",
    "            sample_id, set()\n",
    "        ).add(ts_threads.population(ts_threads.node(node).population).metadata[\"breed\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aedd66a5",
   "metadata": {},
   "source": [
    "Test that sample and breed mapping is correct:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "62dc72e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test that sample and breed mapping is correct:\n",
    "assert all(\n",
    "    sample_info.loc[sample_info[\"individual\"] == sample_id, \"population\"].values[0] in breed_set\n",
    "    for sample_id, breed_set in sample2breed.items()\n",
    "), \"Sample to breed mapping is incorrect!\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "915cddfe",
   "metadata": {},
   "source": [
    "Now test that the genotypes are consistent with sample names: using VCF as a reference,\n",
    "read the `focal` sample generated through the pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5b24c15c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['tsk_105', 'tsk_106', 'tsk_107', 'tsk_108', 'tsk_109']\n"
     ]
    }
   ],
   "source": [
    "vcf_file = get_project_dir() / \"results-threads/toInfer/focal/ts300I2k.1.vcf.gz\"\n",
    "with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "    print(vcf_reader.samples[:5])  # show first 5 sample names in VCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aafb37a4",
   "metadata": {},
   "source": [
    "Check that the number of variant is the same:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2356a514",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N. of variants in VCF: 2000\n",
      "N. of variants TreeSequence: 1998\n",
      "✗ Warning: VCF has 2000 variants, TreeSequence has 1998\n"
     ]
    }
   ],
   "source": [
    "# count number of variants in VCF file\n",
    "num_variants = sum(1 for _ in cyvcf2.VCF(vcf_file))\n",
    "print(f\"N. of variants in VCF: {num_variants}\")\n",
    "\n",
    "# count number of variants in TS file\n",
    "num_ts_variants = ts_threads.num_sites\n",
    "print(f\"N. of variants TreeSequence: {num_ts_variants}\")\n",
    "\n",
    "# check variant size\n",
    "if num_variants == num_ts_variants:\n",
    "    print(f\"✓ N. of variants matches (between TS and VCF)!\")\n",
    "else:\n",
    "    print(f\"✗ Warning: VCF has {num_variants} variants, TreeSequence has {num_ts_variants}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a153b0ed",
   "metadata": {},
   "source": [
    "Ok, positions are not the same: for what I saw, first position in `ts_threads` is 0,\n",
    "so let's read first position on vcf file and determine the offset of *TS* object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "516f2c16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[  8568  11982  54502  99207  99302 110491 132927 136059 144274 177369]\n",
      "[  8568  11982  54502  99207  99302 110491 132927 136059 144274 177369]\n",
      "[  8568  11982  54502  99207  99302 110491 132927 136059 144274 177369]\n"
     ]
    }
   ],
   "source": [
    "with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "    variant = next(vcf_reader)\n",
    "    offset = variant.POS\n",
    "\n",
    "# translate the TS positions\n",
    "ts_positions = np.array([int(site.position) + offset for site in ts_threads.sites()], dtype=np.int64)\n",
    "print(ts_positions[:10])  # show first 10 positions\n",
    "\n",
    "# read positions from VCF\n",
    "with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "    vcf_positions = np.array([variant.POS for variant in vcf_reader])\n",
    "print(vcf_positions[:10])  # show first 10 positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6aa324d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TreeSequence positions: 1998 variants\n",
      "VCF positions: 2000 variants\n",
      "\n",
      "Extracting variant positions from VCF...\n",
      "\n",
      "Variants in VCF but missing in TreeSequence: 2\n",
      "\n",
      "Variants in VCF but missing in TreeSequence: 2\n"
     ]
    }
   ],
   "source": [
    "# Find which variants are missing from the TreeSequence\n",
    "# Get positions from both VCF and TreeSequence\n",
    "print(f\"TreeSequence positions: {len(ts_positions)} variants\")\n",
    "print(f\"VCF positions: {len(vcf_positions)} variants\")\n",
    "\n",
    "print(\"\\nExtracting variant positions from VCF...\")\n",
    "\n",
    "# collect information on variants\n",
    "vcf_variants_info = []\n",
    "\n",
    "with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "    for variant in vcf_reader:\n",
    "        vcf_variants_info.append({\n",
    "            'chrom': variant.CHROM,\n",
    "            'pos': variant.POS,\n",
    "            'ref': variant.REF,\n",
    "            'alt': ','.join(variant.ALT) if variant.ALT else '',\n",
    "            'qual': variant.QUAL,\n",
    "            'filter': variant.FILTER if variant.FILTER else 'PASS',\n",
    "        })\n",
    "\n",
    "# create a dataframe from variant information\n",
    "vcf_variants_df = pd.DataFrame(vcf_variants_info)\n",
    "\n",
    "# Find missing positions (in VCF but not in TreeSequence)\n",
    "missing_in_ts = np.setdiff1d(vcf_positions, ts_positions)\n",
    "print(f\"\\nVariants in VCF but missing in TreeSequence: {len(missing_in_ts)}\")\n",
    "\n",
    "if len(missing_in_ts) > 0:\n",
    "    # filter out missing variants\n",
    "    missing_variants_df = vcf_variants_df[vcf_variants_df['pos'].isin(missing_in_ts)].copy()\n",
    "\n",
    "    genotype_stats = []\n",
    "\n",
    "    with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "        for variant in vcf_reader:\n",
    "            if variant.POS in missing_in_ts:\n",
    "                gts = np.array([gt[:2] for gt in variant.genotypes])\n",
    "                unique_alleles = np.unique(gts[gts >= 0])  # Escludere missing (-1)\n",
    "\n",
    "                genotype_stats.append({\n",
    "                    'pos': variant.POS,\n",
    "                    'unique_alleles': len(unique_alleles),\n",
    "                    'is_monomorphic': len(unique_alleles) == 1,\n",
    "                    'missing_gts': np.sum(gts == -1),\n",
    "                    'total_gts': gts.size,\n",
    "                    'allele_counts': dict(zip(*np.unique(gts.flatten(), return_counts=True)))\n",
    "                })\n",
    "\n",
    "    genotype_stats_df = pd.DataFrame(genotype_stats)\n",
    "\n",
    "    # join information\n",
    "    missing_variants_full = missing_variants_df.merge(genotype_stats_df, on='pos')\n",
    "\n",
    "    # uncomment to save data\n",
    "    # missing_variants_full.to_csv('missing_variants.csv', index=False)\n",
    "\n",
    "else:\n",
    "    print(\"\\n✓ All VCF variants are present in TreeSequence\")\n",
    "\n",
    "# Also check if TreeSequence has any variants not in VCF (unlikely but possible)\n",
    "extra_in_ts = np.setdiff1d(ts_positions, vcf_positions)\n",
    "if len(extra_in_ts) > 0:\n",
    "    extra_df = pd.DataFrame({'pos': extra_in_ts})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc8cb2a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chrom",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "pos",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "ref",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "alt",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "qual",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "filter",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "unique_alleles",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "is_monomorphic",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "missing_gts",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "total_gts",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "allele_counts",
         "rawType": "object",
         "type": "unknown"
        }
       ],
       "ref": "afe8edfd-e4a0-4499-8672-00e6f73b4cd7",
       "rows": [
        [
         "0",
         "1",
         "45384357",
         "C",
         "T",
         null,
         "PASS",
         "1",
         "True",
         "0",
         "3200",
         "{np.int64(0): np.int64(3200)}"
        ],
        [
         "1",
         "1",
         "63257441",
         "C",
         "T,G",
         null,
         "PASS",
         "3",
         "False",
         "0",
         "3200",
         "{np.int64(0): np.int64(3145), np.int64(1): np.int64(7), np.int64(2): np.int64(48)}"
        ]
       ],
       "shape": {
        "columns": 11,
        "rows": 2
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>unique_alleles</th>\n",
       "      <th>is_monomorphic</th>\n",
       "      <th>missing_gts</th>\n",
       "      <th>total_gts</th>\n",
       "      <th>allele_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>45384357</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>3200</td>\n",
       "      <td>{0: 3200}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>63257441</td>\n",
       "      <td>C</td>\n",
       "      <td>T,G</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>3200</td>\n",
       "      <td>{0: 3145, 1: 7, 2: 48}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom       pos ref  alt  qual filter  unique_alleles  is_monomorphic  \\\n",
       "0     1  45384357   C    T  None   PASS               1            True   \n",
       "1     1  63257441   C  T,G  None   PASS               3           False   \n",
       "\n",
       "   missing_gts  total_gts           allele_counts  \n",
       "0            0       3200               {0: 3200}  \n",
       "1            0       3200  {0: 3145, 1: 7, 2: 48}  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_variants_full"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d92f48d",
   "metadata": {},
   "source": [
    "So the missing variants are one monomorphic SNP (no alternates) in position `45384357` and\n",
    "one biallelic SNP in position `63257441` (filtered out by the `BCFTOOLS_BIALLELIC` step).\n",
    "Let's track the position to skip:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e32bd600",
   "metadata": {},
   "outputs": [],
   "source": [
    "skip_positions = missing_variants_full[\"pos\"].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5fd06e3",
   "metadata": {},
   "source": [
    "## Extract TreeSequences Genotypes\n",
    "\n",
    "Collect the full genotype matrix from TS object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d2dc5dd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotype matrix from TS: (1998, 3200)\n",
      "Number of sites: 1998\n",
      "Number of samples: 3200\n",
      "\n",
      "Subsetting first 5 objects in both dimensions\n",
      "[[0 0 0 0 0]\n",
      " [0 0 0 0 1]\n",
      " [0 0 0 1 1]\n",
      " [0 0 0 1 0]\n",
      " [0 0 0 1 0]]\n"
     ]
    }
   ],
   "source": [
    "# collect a matrix of (num_sites, num_samples)\n",
    "ts_genotype_matrix = ts_threads.genotype_matrix()\n",
    "print(f\"Genotype matrix from TS: {ts_genotype_matrix.shape}\")\n",
    "print(f\"Number of sites: {ts_genotype_matrix.shape[0]}\")\n",
    "print(f\"Number of samples: {ts_genotype_matrix.shape[1]}\")\n",
    "print(f\"\\nSubsetting first 5 objects in both dimensions\")\n",
    "print(ts_genotype_matrix[:5, :5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb6b5e5b",
   "metadata": {},
   "source": [
    "## Extract VCF genotypes\n",
    "\n",
    "Collect the full genotype matrix from VCF (remember to skip filtered variants):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2ad2fa14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Extracting genotypes from VCF...\n",
      "\n",
      "VCF genotype matrix created:\n",
      "  Shape: (1998, 3200)\n",
      "  Dtype: int8\n",
      "  Number of variants: 1998\n",
      "  Number of samples: 3200\n",
      "\n",
      "Example - first 5 genotypes of the first sample:\n",
      "[[0 0 0 0 0]\n",
      " [0 0 0 0 1]\n",
      " [0 0 0 1 1]\n",
      " [0 0 0 1 0]\n",
      " [0 0 0 1 0]]\n",
      "\n",
      "VCF genotype matrix created:\n",
      "  Shape: (1998, 3200)\n",
      "  Dtype: int8\n",
      "  Number of variants: 1998\n",
      "  Number of samples: 3200\n",
      "\n",
      "Example - first 5 genotypes of the first sample:\n",
      "[[0 0 0 0 0]\n",
      " [0 0 0 0 1]\n",
      " [0 0 0 1 1]\n",
      " [0 0 0 1 0]\n",
      " [0 0 0 1 0]]\n"
     ]
    }
   ],
   "source": [
    "with cyvcf2.VCF(vcf_file) as vcf_reader:\n",
    "    print(\"\\nExtracting genotypes from VCF...\")\n",
    "\n",
    "    genotype_list = []\n",
    "\n",
    "    for i, variant in enumerate(vcf_reader):\n",
    "        if variant.POS in skip_positions:\n",
    "            continue  # those variants are missing in TS\n",
    "\n",
    "        # Extract alleles as a numpy array directly\n",
    "        # variant.genotypes is a list, we take only the first 2 elements (alleles)\n",
    "        # Unnest genotypes: create a (num_samples * 2,) flat array for this variant\n",
    "        genotypes_array = np.array([gt[:2] for gt in variant.genotypes], dtype=np.int8)\n",
    "        genotype_list.append(genotypes_array.flatten())\n",
    "\n",
    "        if (i + 1) % 10000 == 0:\n",
    "            print(f\"  Processed {i + 1} variants...\")\n",
    "\n",
    "    # Convert to a 3D numpy array: dimensions (num_variants, num_samples, 2)\n",
    "    vcf_genotype_matrix = np.array(genotype_list, dtype=np.int8)\n",
    "\n",
    "    print(f\"\\nVCF genotype matrix created:\")\n",
    "    print(f\"  Shape: {vcf_genotype_matrix.shape}\")\n",
    "    print(f\"  Dtype: {vcf_genotype_matrix.dtype}\")\n",
    "    print(f\"  Number of variants: {vcf_genotype_matrix.shape[0]}\")\n",
    "    print(f\"  Number of samples: {vcf_genotype_matrix.shape[1]}\")\n",
    "\n",
    "    print(f\"\\nExample - first 5 genotypes of the first sample:\")\n",
    "    print(vcf_genotype_matrix[:5, :5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "494b78f4",
   "metadata": {},
   "source": [
    "## Compare genotype matrices\n",
    "\n",
    "Compare the two genotype matrices:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "74cc3f3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Matrices have the same shape: (1998, 3200)\n",
      "\n",
      "Global result:\n",
      "  Sites with matching genotypes: 1998/1998 (100.00%)\n",
      "\n",
      "✓ All matrices match perfectly!\n",
      "\n",
      "Samples with all matching genotypes: 3200/3200\n"
     ]
    }
   ],
   "source": [
    "# Check matching between VCF and TreeSequence genotype matrices: test for shape and values\n",
    "if vcf_genotype_matrix.shape == ts_genotype_matrix.shape:\n",
    "    print(f\"✓ Matrices have the same shape: {vcf_genotype_matrix.shape}\")\n",
    "\n",
    "    # Both matrices have shape (num_sites, num_samples * 2 intended as haploid individuals)\n",
    "    # test for site-wise matches\n",
    "    matches = np.all(vcf_genotype_matrix == ts_genotype_matrix, axis=1)\n",
    "    num_matches = np.sum(matches)\n",
    "    num_total = len(matches)\n",
    "\n",
    "    print(f\"\\nGlobal result:\")\n",
    "    print(f\"  Sites with matching genotypes: {num_matches}/{num_total} ({100*num_matches/num_total:.2f}%)\")\n",
    "\n",
    "    if num_matches != num_total:\n",
    "        # Find sites with mismatches\n",
    "        mismatches = np.where(~matches)[0]\n",
    "        print(f\"\\n  Number of sites with mismatches: {len(mismatches)}\")\n",
    "        print(f\"  First 5 sites with mismatches: {mismatches[:5]}\")\n",
    "\n",
    "        # Analyze an example of mismatch\n",
    "        if len(mismatches) > 0:\n",
    "            site_idx = mismatches[0]\n",
    "            print(f\"\\n  Example - Site {site_idx}:\")\n",
    "\n",
    "            # Find which samples have mismatches\n",
    "            site_matches = np.all(vcf_genotype_matrix[site_idx] == ts_genotype_matrix[site_idx], axis=1)\n",
    "            discordant_samples = np.where(~site_matches)[0]\n",
    "\n",
    "            print(f\"    Samples with mismatches: {len(discordant_samples)}\")\n",
    "            if len(discordant_samples) > 0:\n",
    "                sample_idx = discordant_samples[0]\n",
    "                print(f\"    Sample {sample_idx}:\")\n",
    "                print(f\"      VCF:          {vcf_genotype_matrix[site_idx, sample_idx]}\")\n",
    "                print(f\"      TreeSequence: {ts_genotype_matrix[site_idx, sample_idx]}\")\n",
    "    else:\n",
    "        print(\"\\n✓ All matrices match perfectly!\")\n",
    "\n",
    "    # Statistics per sample\n",
    "    # Per-sample match stats (all sites must match for a sample to be \"perfect\")\n",
    "    sample_matches = np.all(vcf_genotype_matrix == ts_genotype_matrix, axis=0)\n",
    "    num_perfect_samples = int(np.sum(sample_matches))\n",
    "\n",
    "    print(f\"\\nSamples with all matching genotypes: {num_perfect_samples}/{ts_threads.num_samples}\")\n",
    "\n",
    "    if num_perfect_samples != ts_threads.num_samples:\n",
    "        mismatched_samples = np.where(~sample_matches)[0]\n",
    "        print(f\"  Number of samples with mismatches: {len(mismatched_samples)}\")\n",
    "        print(f\"  First 10 samples with mismatches: {mismatched_samples[:10].tolist()}\")\n",
    "else:\n",
    "    print(f\"✗ Matrices have different shapes!\")\n",
    "    print(f\"  VCF:          {vcf_genotype_matrix.shape}\")\n",
    "    print(f\"  TreeSequence: {ts_genotype_matrix.shape}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tskitetude-hh-GIRXc-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
