{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TSINFER tutorial\n",
    "\n",
    "`tsinfer` has changed a lot in after this whole project started. Starting from\n",
    "version `0.4.0` the `SampleData` class has beed deprecated. So tutorials change a \n",
    "lot, and the process of inference is completely different\n",
    "\n",
    "## Toy example\n",
    "\n",
    "`tsinfer` takes data into [zarr](https://zarr.readthedocs.io/en/stable/) format, \n",
    "with *phased variants* in [VCF zarr](https://sgkit-dev.github.io/bio2zarr/intro.html)\n",
    "format. I've downloaded the \n",
    "[example data](https://github.com/tskit-dev/tsinfer/tree/main/docs/_static/example_data.vcz)\n",
    "from GitHub:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import zarr\n",
    "import tsinfer\n",
    "import pyfaidx\n",
    "\n",
    "from tskitetude import get_data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_zarr = zarr.open(get_data_dir() / \"example_data.vcz\")\n",
    "vcf_zarr.tree()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how I can access to a zarr *group* and display something similar to the \n",
    "tutorial:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# display genotypes for the first few samples and sites\n",
    "positions = vcf_zarr['variant_position'][:]\n",
    "alleles = vcf_zarr['variant_allele'][:]\n",
    "genotypes = vcf_zarr['call_genotype'][:]\n",
    "\n",
    "# Number of samples to display (first 3 diploid samples)\n",
    "num_samples = 3\n",
    "num_sites = min(8, len(positions))  # First 8 sites or fewer if not available\n",
    "\n",
    "# Header - Diploid sample id\n",
    "print(\"Diploid sample id: \", end=\"\")\n",
    "for i in range(num_samples):\n",
    "    print(f\"    {i:<10}\", end=\"\")\n",
    "print()\n",
    "\n",
    "# Header - Genome for sample\n",
    "print(\"Genome for sample: \", end=\"\")\n",
    "for i in range(num_samples):\n",
    "    print(f\" {i*2:<5} {i*2+1:<7}\", end=\"\")\n",
    "print()\n",
    "\n",
    "# Separator line\n",
    "print(\"-\" * (19 + num_samples * 14))\n",
    "\n",
    "# Site data\n",
    "for site_idx in range(num_sites):\n",
    "    pos = positions[site_idx]\n",
    "    site_alleles = alleles[site_idx]\n",
    "\n",
    "    print(f\"      position {pos:<2}: \", end=\"\")\n",
    "\n",
    "    for sample_idx in range(num_samples):\n",
    "        # Genotypes for this sample (diploid, so 2 values)\n",
    "        g0 = genotypes[site_idx, sample_idx, 0]\n",
    "        g1 = genotypes[site_idx, sample_idx, 1]\n",
    "\n",
    "        # Convert indices to alleles\n",
    "        allele0 = site_alleles[g0].decode() if g0 >= 0 else \".\"\n",
    "        allele1 = site_alleles[g1].decode() if g1 >= 0 else \".\"\n",
    "\n",
    "        print(f\" {allele0:<5} {allele1:<7}\", end=\"\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "The last site, at position 95, is an indel (insertion or deletion). Indels can be used for inference as long as the indel does not overlap with other variants, only two alleles exist, and the ancestral state is known.\n",
    ":::\n",
    "\n",
    "## VariantData\n",
    "\n",
    "The `VariantData` is the new object used by `tsinfer` to infer *trees*. It requires\n",
    "two parameters, first is a *zarr group* or a path to a *zarr file*. The second parameter\n",
    "is the `ancestral_state` for each site used in reference. If the vcf file has\n",
    "a *INFO* filed where the ancestral state is specified, the *zarr group* will have\n",
    "such key, so it can be provided to `VariantData`. In this example, we don't \n",
    "have such field, and we can provide it trough a reference sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reader = pyfaidx.Fasta(get_data_dir() / \"example_ancestral_state.fa\")\n",
    "ancestral_str = str(reader[\"chr1\"])\n",
    "ancestral_str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the code above, we have simply read a reference sequence into a `str` object.\n",
    "This sequence is a fake object, it has letters representing ancestral alleles. All\n",
    "the other positions are simply marked with \"N\" (unknown). Since the position in \n",
    "VCF are 1 based and the sequence is 0 based, we simply add a *letter* at the \n",
    "beginning of the string to use the same indexes: with `tsinfer.add_ancestral_state_array`\n",
    "the ancestral state is add to *zarr group*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestral_str = \"X\" + ancestral_str\n",
    "\n",
    "# Add ancestral state only if not already present (idempotent operation)\n",
    "if \"ancestral_state\" not in vcf_zarr:\n",
    "    tsinfer.add_ancestral_state_array(vcf_zarr, ancestral_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"ancestral_state\" in vcf_zarr:\n",
    "    print(vcf_zarr[\"ancestral_state\"][:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdata = tsinfer.VariantData(get_data_dir() / \"example_data.vcz\", ancestral_state=\"ancestral_state\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ":::{note}\n",
    "Because the ancestral state at position 85 does not match any of the alleles at \n",
    "that site, a warning has been given that the ancestral state will be considered \n",
    "unknown. As a consequence, although it will appear in the inferred tree sequence, \n",
    "the site at position 85 will be treated as a \"_noninference_\" site.\n",
    ":::\n",
    "\n",
    "This is not the only way to provide ancestral state information. You can also\n",
    "add the reference (\"REF\") allele as the ancestral state, taking advantage of the \n",
    "fact that the `.vcz` format always stores the zeroth allele as the REF, \n",
    "suggesting the following syntax:\n",
    "\n",
    "```python\n",
    "vdata = tsinfer.VariantData(vcf_zarr, ancestral_state=vcf_zarr[\"variant_allele\"][:, 0])\n",
    "```\n",
    "\n",
    "After creating a `VariantData` we can calculate a *TreeSequence* object with `tsinfer`:\n",
    "each sample in the original `.vcz` file will correspond to an individual in the \n",
    "resulting *TreeSequence*. With 3 diploid individuals, we will have 6 *TreeSequence*\n",
    "haploid individuals:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inferred_ts = tsinfer.infer(vdata)\n",
    "print(f\"Inferred a genetic genealogy for {inferred_ts.num_samples} (haploid) genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's display the *TreeSequence* object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inferred_ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Non-inference sites\n",
    "\n",
    "Certain sites are not used by `tsinfer` for inferring the genealogy. These **noninference sites** are nevertheless included in the final tree sequence, but with their mutations placed by parsimony. Such sites include:\n",
    "\n",
    "- **Non-variable (fixed) sites**, e.g. the site at position 71 above\n",
    "- **Singleton sites**, where only one genome has the derived allele, e.g. the site at position 75 above\n",
    "- **Sites where the ancestral allele is unknown**, e.g. site 85 (see above)\n",
    "- **Multiallelic sites**, with more than 2 alleles (but see [here](https://tskit.dev/tsinfer/docs/stable/tutorial.html#multiallelic-sites) for a workaround)\n",
    "\n",
    "Additional sites can be deliberately flagged as not-for-use in inference, for example if their genotypes or ancestral states are deemed unreliable, via the `exclude_positions` parameter when running the inference step of `tsinfer`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "import numpy as np\n",
    "import tsinfer\n",
    "import cyvcf2\n",
    "import json\n",
    "import tsdate\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "from tskit import MISSING_DATA\n",
    "\n",
    "from tskitetude import get_data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with tsinfer.SampleData(sequence_length=6) as sample_data:\n",
    "    sample_data.add_site(0, [0, 1, 0, 0, 0], [\"A\", \"T\"], ancestral_allele=0)\n",
    "    sample_data.add_site(1, [0, 0, 0, 1, 1], [\"G\", \"C\"], ancestral_allele=0)\n",
    "    sample_data.add_site(2, [0, 1, 1, 0, 0], [\"C\", \"A\"], ancestral_allele=0)\n",
    "    sample_data.add_site(3, [0, 1, 1, 0, 0], [\"G\", \"C\"], ancestral_allele=MISSING_DATA)\n",
    "    sample_data.add_site(4, [0, 0, 0, 1, 1], [\"A\", \"C\"], ancestral_allele=0)\n",
    "    sample_data.add_site(5, [0, 1, 2, 0, 0], [\"T\", \"G\", \"C\"], ancestral_allele=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`tsinfer.Sampledata` is the object required for inferring a `tstree` object. Using \n",
    "the `add_site()` method a can add information for each SNP respectively. The first\n",
    "argument is the *SNP position*: here for simplicity we track SNP in positional order\n",
    "but it can be any positive value (even float). The only requirement is that this \n",
    "position should be unique and added in increasing order. The 2nd argument is for \n",
    "the *genotypes* of each sample in this position: is and index of the allele I can \n",
    "find in the 3rd argument. If I have a missing data, I need to use the `tskit.MISSING_DATA`\n",
    "The last argument is the index of the ancestral allele. Not all the sites are used\n",
    "to infer the *tree* object: sites with missing data or ancestral alleles or sites with\n",
    "more than 2 genotypes are not considered by will be modeled in the resulting tree.\n",
    "Once we have the `SampleData` instance, we can infer a `tstree` object using\n",
    "`tsinfer.infer`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = tsinfer.infer(sample_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This `ts` object is a full *Tree Sequence* object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This *Tree sequence* object can be analyzed as usual:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"==Haplotypes==\")\n",
    "for sample_id, h in enumerate(ts.haplotypes()):\n",
    "    print(sample_id, h, sep=\"\\t\")\n",
    "ts.draw_svg(y_axis=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If I understand correctly, `tsinfer` can impute missing data (check this). For the\n",
    "data I put, there's a *root* note with three *childs*: this is also known as *polytomy*.\n",
    "Every *internal* node represent an ancestral sequence, By default, the time of those\n",
    "nodes is not measured in years or generations, but is the frequency of the shared\n",
    "derived alleles on which the ancestral sequence is based. This is why the time is\n",
    "*uncalibrated* in the graph above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extra code to label and order the tips alphabetically rather than numerically\n",
    "labels = {i: string.ascii_lowercase[i] for i in range(ts.num_nodes)}\n",
    "genome_order = [n for n in ts.first().nodes(order=\"minlex_postorder\") if ts.node(n).is_sample()]\n",
    "labels.update({n: labels[i] for i, n in enumerate(genome_order)})\n",
    "style1 = (\n",
    "    \".node:not(.sample) > .sym, .node:not(.sample) > .lab {visibility: hidden;}\"\n",
    "    \".mut {font-size: 12px} .y-axis .tick .lab {font-size: 85%}\")\n",
    "sz = (800, 250)  # size of the plot, slightly larger than the default\n",
    "\n",
    "# ticks = [0, 5000, 10000, 15000, 20000]\n",
    "# get max generations time:\n",
    "max_time = ts.node(ts.get_num_nodes() - 1).time\n",
    "ticks = np.linspace(0, max_time, 5)\n",
    "ts.draw_svg(\n",
    "    size=sz, node_labels=labels, style=style1, y_label=\"Time ago (uncalibrated)\",\n",
    "    y_axis=True, y_ticks=ticks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inferring dates\n",
    "\n",
    "To infer *dates* we can use `tsdate.date` with default parameter and by \n",
    "specifying *Effective population sizes* and *mutation rate*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dated_ts = tsdate.date(ts, method=\"inside_outside\", mutation_rate=1e-8, population_size=1e4, progress=True)\n",
    "dated_ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dated_ts.draw_svg(y_axis=True, size=(800, 250))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data example\n",
    "\n",
    "This is the [Data example](https://tskit.dev/tsinfer/docs/stable/tutorial.html#data-example)\n",
    "part of the tutorial:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_diploid_sites(vcf, samples):\n",
    "    \"\"\"\n",
    "    Read the sites in the vcf and add them to the samples object.\n",
    "    \"\"\"\n",
    "    # You may want to change the following line, e.g. here we allow\n",
    "    # \"*\" (a spanning deletion) to be a valid allele state\n",
    "    allele_chars = set(\"ATGCatgc*\")\n",
    "    pos = 0\n",
    "    progressbar = tqdm(total=samples.sequence_length, desc=\"Read VCF\", unit='bp')\n",
    "\n",
    "    for variant in vcf:  # Loop over variants, each assumed at a unique site\n",
    "        progressbar.update(variant.POS - pos)\n",
    "\n",
    "        if pos == variant.POS:\n",
    "            print(f\"Duplicate entries at position {pos}, ignoring all but the first\")\n",
    "            continue\n",
    "\n",
    "        else:\n",
    "            pos = variant.POS\n",
    "\n",
    "        if any([not phased for _, _, phased in variant.genotypes]):\n",
    "            raise ValueError(\"Unphased genotypes for variant at position\", pos)\n",
    "\n",
    "        alleles = [variant.REF.upper()] + [v.upper() for v in variant.ALT]\n",
    "        ancestral = variant.INFO.get(\"AA\", \".\")  # \".\" means unknown\n",
    "\n",
    "        # some VCFs (e.g. from 1000G) have many values in the AA field: take the 1st\n",
    "        ancestral = ancestral.split(\"|\")[0].upper()\n",
    "\n",
    "        if ancestral == \".\" or ancestral == \"\":\n",
    "            ancestral_allele = MISSING_DATA\n",
    "            # alternatively, you could specify `ancestral = variant.REF.upper()`\n",
    "\n",
    "        else:\n",
    "            ancestral_allele = alleles.index(ancestral)\n",
    "\n",
    "        # Check we have ATCG alleles\n",
    "        for a in alleles:\n",
    "            if len(set(a) - allele_chars) > 0:\n",
    "                print(f\"Ignoring site at pos {pos}: allele {a} not in {allele_chars}\")\n",
    "                continue\n",
    "\n",
    "        # Map original allele indexes to their indexes in the new alleles list.\n",
    "        genotypes = [g for row in variant.genotypes for g in row[0:2]]\n",
    "        samples.add_site(pos, genotypes, alleles, ancestral_allele=ancestral_allele)\n",
    "\n",
    "\n",
    "def chromosome_length(vcf):\n",
    "    assert len(vcf.seqlens) == 1\n",
    "    return vcf.seqlens[0]\n",
    "\n",
    "\n",
    "# NB: could also read from an online version by setting vcf_location to\n",
    "# \"https://github.com/tskit-dev/tsinfer/raw/main/docs/_static/P_dom_chr24_phased.vcf.gz\"\n",
    "vcf_location =  get_data_dir() / \"P_dom_chr24_phased.vcf.gz\"\n",
    "samples_location = get_data_dir() / \"P_dom_chr24_phased.samples\"\n",
    "\n",
    "vcf = cyvcf2.VCF(vcf_location)\n",
    "\n",
    "with tsinfer.SampleData(\n",
    "    path=str(samples_location), sequence_length=chromosome_length(vcf)\n",
    ") as samples:\n",
    "    add_diploid_sites(vcf, samples)\n",
    "\n",
    "print(\n",
    "    \"Sample file created for {} samples \".format(samples.num_samples)\n",
    "    + \"({} individuals) \".format(samples.num_individuals)\n",
    "    + \"with {} variable sites.\".format(samples.num_sites),\n",
    "    flush=True,\n",
    ")\n",
    "\n",
    "# Do the inference\n",
    "ts = tsinfer.infer(samples)\n",
    "print(\n",
    "    \"Inferred tree sequence: {} trees over {} Mb ({} edges)\".format(\n",
    "        ts.num_trees, ts.sequence_length / 1e6, ts.num_edges\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's also a parallel version of this `add_diploid_sites` [here](https://github.com/tskit-dev/tsinfer/issues/277#issuecomment-652024871).\n",
    "Well, until now I added 20 different individuals (with a single chromosome) instead \n",
    "of adding 10 diploid individuals. I can change something to add more chromosomes \n",
    "to the same individual, and even add other meta information to the three:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_populations(vcf, samples):\n",
    "    \"\"\"\n",
    "    Add tsinfer Population objects and returns a list of IDs corresponding to the VCF samples.\n",
    "    \"\"\"\n",
    "\n",
    "    # In this VCF, the first letter of the sample name refers to the population\n",
    "    samples_first_letter = [sample_name[0] for sample_name in vcf.samples]\n",
    "\n",
    "    pop_lookup = {}\n",
    "    pop_lookup[\"8\"] = samples.add_population(metadata={\"country\": \"Norway\"})\n",
    "    pop_lookup[\"F\"] = samples.add_population(metadata={\"country\": \"France\"})\n",
    "\n",
    "    return [pop_lookup[first_letter] for first_letter in samples_first_letter]\n",
    "\n",
    "\n",
    "def add_diploid_individuals(vcf, samples, populations):\n",
    "    for name, population in zip(vcf.samples, populations):\n",
    "        samples.add_individual(ploidy=2, metadata={\"name\": name}, population=population)\n",
    "\n",
    "\n",
    "# Repeat as previously but add both populations and individuals\n",
    "vcf_location =  get_data_dir() / \"P_dom_chr24_phased.vcf.gz\"\n",
    "samples_location = get_data_dir() / \"P_dom_chr24_phased.samples\"\n",
    "\n",
    "vcf = cyvcf2.VCF(vcf_location)\n",
    "with tsinfer.SampleData(\n",
    "        path=str(samples_location), sequence_length=chromosome_length(vcf)\n",
    "        ) as samples:\n",
    "    populations = add_populations(vcf, samples)\n",
    "    add_diploid_individuals(vcf, samples, populations)\n",
    "    add_diploid_sites(vcf, samples)\n",
    "\n",
    "print(\n",
    "    \"Sample file created for {} samples \".format(samples.num_samples)\n",
    "    + \"({} individuals) \".format(samples.num_individuals)\n",
    "    + \"with {} variable sites.\".format(samples.num_sites),\n",
    "    flush=True,\n",
    ")\n",
    "\n",
    "# Do the inference\n",
    "sparrow_ts = tsinfer.infer(samples)\n",
    "\n",
    "print(\n",
    "    \"Inferred tree sequence `{}`: {} trees over {} Mb\".format(\n",
    "        \"sparrow_ts\", sparrow_ts.num_trees, sparrow_ts.sequence_length / 1e6\n",
    "    )\n",
    ")\n",
    "# Check the metadata\n",
    "for sample_node_id in sparrow_ts.samples():\n",
    "    individual_id = sparrow_ts.node(sample_node_id).individual\n",
    "    population_id = sparrow_ts.node(sample_node_id).population\n",
    "    print(\n",
    "        \"Node\",\n",
    "        sample_node_id,\n",
    "        \"labels a chr24 sampled from individual\",\n",
    "        json.loads(sparrow_ts.individual(individual_id).metadata),\n",
    "        \"in\",\n",
    "        json.loads(sparrow_ts.population(population_id).metadata)[\"country\"],\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis\n",
    "\n",
    "Now analyses can be done with `tskit` libraries. I can't show the full *tree sequences*\n",
    "for this object, I can focus to a segment however:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colours = {\"Norway\": \"red\", \"France\": \"blue\"}\n",
    "colours_for_node = {}\n",
    "\n",
    "for n in sparrow_ts.samples():\n",
    "    population_data = sparrow_ts.population(sparrow_ts.node(n).population)\n",
    "    colours_for_node[n] = colours[json.loads(population_data.metadata)[\"country\"]]\n",
    "\n",
    "individual_for_node = {}\n",
    "for n in sparrow_ts.samples():\n",
    "    individual_data = sparrow_ts.individual(sparrow_ts.node(n).individual)\n",
    "    individual_for_node[n] = json.loads(individual_data.metadata)[\"name\"]\n",
    "\n",
    "tree = sparrow_ts.at(1e6)\n",
    "tree.draw(\n",
    "    height=700,\n",
    "    width=1200,\n",
    "    node_labels=individual_for_node,\n",
    "    node_colours=colours_for_node,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tskitetude-QRcCziD4-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
